File: prefix_no_lib/src/main/scala/prefix_app/FourInts.scala
class FourInts(first: Int, second: Int, third: Int, fourth: Int) extends Serializable {

  def getFirst(): Int = this.first
  def getSecond(): Int = this.second
  def getThird(): Int = this.third
  def getFourth(): Int = this.fourth
  def getValue(): Int = {
    this.getSecond() + this.getThird() + this.getFourth()
  }

  override
  def toString: String = {
    this.getFirst().toString + " | " + this.getSecond().toString + " | " + this.getThird + " | " + this.getFourth
  }
}

object FourInts {
  def cmpKey(o: FourInts): Int = {
    o.getFirst()
  }
}

File: hadoop_prefix_no_lib/src/main/java/prefix_app/FourInts.java
package prefix_app;

import java.util.Comparator;
import org.apache.hadoop.io.Text;

public class FourInts {
  public static final String DELIMITER = " ";
  private int first;
  private int second;
  private int third;
  private int fourth;
  private String encodedValues;

  public FourInts(String encodedValues) {
    this.encodedValues = encodedValues;
    String[] values = encodedValues.split(DELIMITER);
    this.first = Integer.parseInt(values[0]);
    this.second = Integer.parseInt(values[1]);
    this.third = Integer.parseInt(values[2]);
    this.fourth = Integer.parseInt(values[3]);
  }

  public FourInts(Text encodedValues) {
    this(encodedValues.toString());
  }

  public long getValue() {
    return this.second + this.third + this.fourth;
  }

  @Override
  public String toString() {
    return this.encodedValues;
  }

  public Text toText() {
    return new Text(this.toString());
  }

  public static class FourIntsComparator implements Comparator<FourInts> {
    @Override
    public int compare(FourInts o1, FourInts o2) {
      return o1.first > o2.first ? 1 : (o1.first < o2.first ? -1 : 0);
    }
  }

  public static Comparator<FourInts> cmp = new FourIntsComparator();
}
